src/arbitrary/traits.rs

```diff
@@ -0,0 +1,312 @@
+/// [`any_with::<A>(args)`]: fn.any_with.html
+/// [`arbitrary`]: fn.arbitrary.html
+/// [`Arbitrary`]: trait.Arbitrary.html
+/// [`Strategy`]: ../strategy/trait.Strategy.html
+pub fn any<A: Arbitrary>() -> StrategyFor<A> {
+    // ^-- We use a shorter name so that turbofish becomes more ergonomic.
+    A::arbitrary()
+}
+
+/// Generates a [`Strategy`] producing [`Arbitrary`] values of `A` with the
+/// given configuration arguments passed in `args`. Unlike [`arbitrary_with`],
+/// it should be used for being explicit on what `A` is.
+/// For clarity, this may be a good idea.
+///
+/// Use this version instead of [`arbitrary_with`] if you want to be clear which
+/// type you want to generate a `Strategy` for, or if you don't have an anchoring
+/// type for type inference to work with.
+///
+/// If you don't want to specify any arguments and instead use the default
+/// behavior, you should use [`any::<A>()`].
+///
+/// # Example
+///
+/// The function can be used as:
+///
+/// ```rust
+/// #[macro_use] extern crate proptest;
+/// use proptest::prelude::any_with;
+/// use proptest::collection::size_range;
+///
+/// proptest! {
+///     fn reverse_reverse_is_identity
+///         (ref vec in any_with::<Vec<u32>>(size_range(1000).lift()))
+///     {
+///         let vec2 = vec.iter().cloned().rev().rev().collect::<Vec<u32>>();
+///         prop_assert_eq!(vec, &vec2);
+///     }
+/// }
+///
+/// fn main() {
+///     reverse_reverse_is_identity();
+/// }
+/// ```
+///
+/// [`any::<A>()`]: fn.any.html
+/// [`arbitrary_with`]: fn.arbitrary_with.html
+/// [`Arbitrary`]: trait.Arbitrary.html
+/// [`Strategy`]: ../strategy/trait.Strategy.html
+pub fn any_with<A: Arbitrary>(args: ParamsFor<A>) -> StrategyFor<A> {
+    // ^-- We use a shorter name so that turbofish becomes more ergonomic.
+    A::arbitrary_with(args)
+}
+
+/// Generates a [`Strategy`] producing [`Arbitrary`] values of `A`.
+/// Works better with type inference than [`any::<A>()`].
+///
+/// With this version, you shouldn't need to specify any of the (many) type
+/// parameters explicitly. This can have a positive effect on type inference.
+/// However, if you want specify `A`, you should use [`any::<A>()`] instead.
+///
+/// For clarity, it is often a good idea to specify the type generated, and
+/// so using [`any::<A>()`] can be a good idea.
+///
+/// If you want to customize how the strategy is generated, use
+/// [`arbitrary_with(args)`] where `args` is of type
+/// `<A as Arbitrary>::Parameters`.
+///
+/// # Example
+///
+/// The function can be used as:
+///
+/// ```rust
+/// extern crate proptest;
+/// use proptest::arbitrary::{arbitrary, StrategyFor};
+///
+/// fn gen_vec_usize() -> StrategyFor<Vec<usize>> {
+///     arbitrary()
+/// }
+///
+/// # fn main() {}
+/// ```
+///
+/// [`arbitrary_with(args)`]: fn.arbitrary_with.html
+/// [`any::<A>()`]: fn.any.html
+/// [`Arbitrary`]: trait.Arbitrary.html
+/// [`Strategy`]: ../strategy/trait.Strategy.html
+pub fn arbitrary<A, S>() -> S
```

Comment:
Here's an example of when there's a difference: https://play.rust-lang.org/?gist=2d255bd0ee2b37d74156b7826ec983fa&version=stable

My understanding is that `arbitrary` provides a bidirectional way to infer `A` from `S` and `S` from `A` because it is eq-constrained that way. But you can't infer `A` from `S` in the case of `any` which is needed to be able to skip specifying `A` in `any::<A>()`.

Reply:
Sounds good, thanks for the example